OEP-50 | initial draft of the hooks extension framework

felipemontoya opened the pull request:
This PR converts the previous proposal held in a google doc to a formal OEP

edx-webhook commented:
Thanks for the pull request, @felipemontoya! I've created [OSPR-5608](https://openedx.atlassian.net/browse/OSPR-5608) to keep track of it in JIRA.

As a core committer in this repo, you can merge this once the pull request is approved per the [core committer reviewer requirements](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529675973/Open+edX+Core+Committers#Reviews-and-Ownership) and according to the agreement with your edX Champion.



<!-- comment:welcome-core-committer -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

felipemontoya commented:
Hello all,
I'm tagging everyone who commented in the previous doc to let you know it is now trying to become an OEP.
@nasthagiri, @ormsbee, @pdpinch, @davidjoy, @bradenmacdonald, @morenol, @diegomillan.

bradenmacdonald commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,227 @@
+lead author, other members of the community, and from one of the largest
+extendable platforms for the web today for many years already, WordPress.
+
+.. _summary of the current options: https://github.com/edx/edx-platform/blob/master/docs/guides/extension_points.rst
```
Just a heads up that this file may be getting moved so the URL will eventually need updating - see https://github.com/edx/edx-developer-docs/pull/37

felipemontoya replied:
Thanks for making me aware of this. I understand that the effort of the developers docs is a bit in a hiatus at the moment. In any case I'll be looking at this to maintain the documentation in the correct place.

bradenmacdonald reviewed (commented):
Thanks @felipemontoya, I am excited to see this.

We had proposed a much simpler and lighter weight mechanism with some similar ideas in https://github.com/edx/edx-django-utils/pull/64 ; I'm mentioning it because it's like a "filter" hook that can be easily declared for small functions. If there's some way that these hooks will cover that same use case, it would be nice. (However, our use for that was something quite insignificant: changing the algorithm for selecting XBlock icons. I'm not sure if hooks are used for things that small?)

pdpinch commented:
very exciting 👍

morenol commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,227 @@
+
+* Most actions will be implemented in plugins.
+* The order of execution for multiple actions listening on a trigger must be
+  respected, and the results of previously executed actions should be made
```
This should be true only for the cases in which the actions run in a sync way (filters). For the async actions this should not apply, right?

felipemontoya replied:
I made it more clear that the order will be kept, but the results will be ignored. This way we will run the pipeline as a whole async, but still keep the order so that others systems can rely on this.

An example would be pipeline with two actions running API calls in an CRM system:

- first action creates the user in the CRM
- second action attaches a piece of info to the user.

These two can run async, but they still need run in this order. Naturally the alternative would be to make one action that does the two things at once, but the whole point of having a pipeline is to be able to divide some work into smaller chunks that complete work by being added together. This again maximizes the flexibility by costing a bit of extra configuration

felipemontoya replied:
With the latest edit async is no longer supported by the framework, so I'll mark this as resolved.

mariajgrimaldi commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,227 @@
+platform must be open for extension following the open-closed principle. This
+OEP specifies a solution using what is commonly referred to as hooking. That is
+the definition of a stable set of places in the code (triggers) where externally
+defined actions (via plugins) can take place.
```
Later in the document, the concept of _filters_ is introduced, but when referring to these external functions that can extend the functionality, only _actions_ are mentioned. That made me think that filters are also actions, if so, it should be indicated that they are. If not, where only _actions_ are mentioned, _or filters_ should be added if appropriate.

felipemontoya replied:
I clarified by better defining actions and filters in the motivation and rationale sections.

felipemontoya commented:
@bradenmacdonald thanks a lot for your comment on the lightweight mechanism created in edx/edx-django-utils#64. We studied it during this last weeks to see how we could better design what we are proposing. We ended up drawing more inspiration from the social-core pipelines for their flexibility.

I don't know what the limit should be for the size or complexity of the hooks framework. For now, I started the proposal with what I though was a comprehensive list of locations that covers the entire lifecycle of a user registering, login in, enrolling, interacting with a course, getting a certificate and loging out. Later we added hooks for the course publication lifecycle. In the future we could add more hooks when the system gains traction and we could also make the exposed triggers somewhat configurable. We have in mind to have a big FEATURE SettingFlag that switches the system on/off. Later on this could be more granular.

felipemontoya commented:
@nizarmah, @bradenmacdonald. You are both candidates for arbiter of this PR. Given that Braden might also want to be a reviewer I was thinking of writing Nizar in.

Do you agree with this?

Also, this has been as a draft since February and the first working document started getting feedback since August last year.

The latest round of implementation discussion in happening in [discuss](https://discuss.openedx.org/t/configuration-for-the-hooks-extension-framework/4527/2). The OEP has been mostly quiet for a while. Would you agree to give a review period of about a month more, to put a review end date by the 23.04.2021?

felipemontoya commented:
@feanil I take from https://github.com/openedx/community-wg/issues/5 that you are reviewing this from edX side. Would agree with this date for the OEP as well? Or am I being too aggressive with the timing here?

We have a proposed implementation that already incorporates much of the feedback and we want to move forward with making a PR for that as well. We'll post the link once is up.

mariajgrimaldi commented:
Here's the draft PR where we're implementing the necessary tooling for the Hooks Extension Framework
[Hooks Extension Framework: Tooling](https://github.com/edx/edx-django-utils/pull/110)

For now, this is a draft PR, but please feel free to leave any comments or feedback!

bradenmacdonald commented:
>  You are both candidates for arbiter of this PR. Given that Braden might also want to be a reviewer I was thinking of writing Nizar in.

I think Nizar would be a great choice! I'm also available if you need a backup :)

nizarmah commented:
Thanks @felipemontoya and @bradenmacdonald for the opportunity 😃 

> The latest round of implementation discussion in happening in discuss. The OEP itself has been mostly quiet for a while. Would you agree to give a review period of about a month more, to put a review end date by the 23.04.2021?

I need to schedule some time tomorrow to go over the discussion on discourse. Sorry about that! 
But I'll do my best to reply within 24 hours 👍🏼

felipemontoya commented:
I added PR https://github.com/edx/edx-platform/pull/27157 with the first view of how the documentation would look.

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,235 @@
+cover the history and all the previous states of the framework. It should
+reflect the latest way of thinking and examples of how to use the framework.
+
+Additionally, a directory in the same location will include individual files
```
[consider] Having individual files for each trigger seems like an implementation detail that may not be needed at the OEP-level, and can evolve over time.

felipemontoya replied:
Removed it in the updated version

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,235 @@
+must consider that:
+
+* Most actions will be implemented in plugins.
+* The order of execution for multiple actions/filters listening on a trigger
```
[clarification] Where and how is the order of execution defined?

felipemontoya replied:
With the separation of django signals for events this would be defined via a config dict in django settings. I left the specifics of the format out since they are more of a implementation detail.

https://github.com/edx/open-edx-proposals/pull/184/files#diff-f28146adea40eb552912482a8b24a19107252beba3e25295413bcd5ef8927628R106-R112

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,235 @@
+  via their results. In this case, the actions called at the time will be called
+  filters.
+* Triggers that will not allow the application flow to be altered by the
+  actions called.
```
[puzzled] This definition of `filters` versus `actions` doesn't match what is written in the Wordpress docs.

Filters do not alter the application flow. Instead, they alter the data that is used within the application flow.
Actions halt the application flow, execute the action, and then return back to the normal flow.

What am I missing in my understanding of this proposal?

felipemontoya replied:
This comment made me clarify a lot of the wording. First I removed "trigger" and "action" from the OEP since it was making it even more complicated. I settled on events and filters (both are kinds of hooks).

I also removed the specific link to the wordpress docs because this proposal is not to copy what Wordpress does, but to be inspired by it.

I clarified the events vs filters part at: https://github.com/edx/open-edx-proposals/pull/184/files#diff-f28146adea40eb552912482a8b24a19107252beba3e25295413bcd5ef8927628R71-R78

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,235 @@
+======================
+
+The only extension point that will be affected by this OEP is the
+REGISTRATION_EXTENSION_FORM. This extension point should still be supported
```
[clarification] Is the proposal to reimplement the current extensibility provided by `REGISTRATION_EXTENSION_FORM` to instead use this new framework?

felipemontoya replied:
I made it clear now. Our suggestion is to keep the current feature until it is clear that it can be implemented using hooks and a migration path has been documented

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,235 @@
+
+* Django Signals was initially considered as the primary mechanism to connect
+triggers and actions, but was ultimately discarded due to the lack of control
+of the execution order when more than one action is listening for a trigger.
```
[seeking more information] Django Signals are the standard Django web framework capability for this type of extension. My initial inclination would be to leverage this capability that already exists in our framework, rather than building and maintaining our own.

What use cases do we have that require a pre-defined execution order?
Also, see my question above since it's not clear how this proposal will satisfy this requirement?

felipemontoya replied:
Now that django signals are back as the mechanism for supporting event hooks, the strict execution order is only there because filters need a way to 'add' or 'reduce' the result of all the functions that were registered.

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,235 @@
+.. _two types of hooks: https://developer.wordpress.org/plugins/hooks/#actions-vs-filters
+
+
+Performance Considerations
```
[request] Can we provide some basis for debugging and troubleshooting performance and functional errors? For example, can we include how one might support distributed tracing so we can track (1) how long each hook's implementation takes and (2) reliable traces/logs through multiple hooks.

felipemontoya replied:
Added a paragraph at the end of this section about that.

nasthagiri reviewed (commented):
Thanks for putting this together.

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,235 @@
+act on it, and return a modified version for the rest of the core-code to use in
+the next steps of execution. We intend to make the two kinds of hooks
+interchangeable such that developers are not limited by the framework but
+instead given maximum flexibility on its usage to attain their extension goals.
```
[strengthening this proposal] It would be great if you could include a few Use Cases (feature-extension examples) for each of the 2 types of hooks. That would bring to light what each hook could enable for the platform.

felipemontoya replied:
Added a few that I think illustrate what I have in mind.
https://github.com/edx/open-edx-proposals/pull/184/files#diff-f28146adea40eb552912482a8b24a19107252beba3e25295413bcd5ef8927628R196-R216

nizarmah commented:
Hello again @felipemontoya! I finally had a chance to go into depth regarding the discussions on [discourse], and the proposal and the comments raised. 

> Would you agree to give a review period of about a month more, to put a review end date by the 23.04.2021?

I think that would be fair or a week after, until 30.04.2021, since some of the discussions on [discourse] have been getting to a resolution, meanwhile, the ones regarding signals/broadcasts and async are still going on, though.

I do have some comments I was considering posting, but I noticed that the proposal hasn't been updated since February 7; therefore, I was thinking of postponing my comments slightly until you have had the chance to update it based on the [discourse] discussion.

[discourse]: https://discuss.openedx.org/t/configuration-for-the-hooks-extension-framework/4527?u=nizar

felipemontoya commented:
I updated the oep with the latest feedback.

In summary:
- clarified the language to use only hooks, events and filters as the main concepts
- made clear that events will use django signals via a new library containing the definitions
- added examples of usage
- included thoughts on async view for django 3.2
- altogether clarified or resolved the open comments (at least I hope they are resolved now)

felipemontoya commented:
@nizarmah @nasthagiri any comments on the current status of this OEP?

Are we ready to merge this latest version?

nizarmah reviewed (approved):
@felipemontoya sorry for the late reply. I have a small comment about a possible typo, but as the arbiter of this OEP, I noticed you've addressed the discussion on the forum and the comments on the pull request in the final changes of this OEP. 

I'll keep an eye for @nasthagiri's comments and make sure things are addressed and provide my approval again on the pull request. 

Thanks for the nice work with the document and the helpful forum updates about the discussions!

felipemontoya commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,288 @@
+
+Since the creation of the plugin framework different initiatives in the
+community have demonstrated success by adopting this pattern. This OEP extends
+the range of possible extensions that the patter can hold sustainably.
```
Good eyes. It is indeed a typo. I'll fix it quick

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,288 @@
+open edX plugins. Initially, this proposal is meant as a way for community
+members to extend the platform, but it could even become a way for the
+installation at edx.org to implement some of its particular business rules in a
+way that is sufficiently separated from the Open edX core.
```
[+1] Yes, completely agree. Once this framework is in place, it would be good to follow-up with technical refactorings of edx.org specific business logic.

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,288 @@
+current receivers of an event.
+
+
+Use Cases
```
[kudos] Excellent examples of use cases. Thx!

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,288 @@
+common case is to be passed data and return the same of somewhat changed, but
+similar data. Other common case would be to receive data and raise an exception
+in accordance with the definition of the hook. Most likely to completely halt
+the process that would happen after the hook.
```
[appreciation] Nice explanation of the difference between events and filters.

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,288 @@
+========================
+
+The reference implementation must be completed before this OEP is given "Final"
+status. It is not necessary for the "accepted" status.
```
[request] hmm... Can we replace "Final"  with "Accepted" and "accepted" with "Provisional"?

```
The reference implementation must be completed before this OEP is given "Accepted" status. It is not necessary for the "Provisional" status.
```

**Rationale:** We don't have any OEPs in the "Final" state and should probably update OEP-1 to exclude that state altogether.  "Accepted"  is synonymous with the community accepting and adopting the proposal.  We _did_ introduce a new state called "Provisional", which means the OEP was reviewed and tentatively accepted while the work is being implemented and piloted.

nasthagiri commented on oeps/oep-0050-hooks-extension-framework.rst:
```diff
@@ -0,0 +1,288 @@
+When possible, hooks should be compatible with the Async Event Messaging
+described in OEP-41.
+In terms of error handling, the implementation should allow the configuration of
+silent failures letting operators know the source of the error.
```
[+1] If I understand this statement correctly, the implementation will have the means to debug issues (logging statements, etc) to help operators understand the root cause of performance and functional issues caused by hooks.

felipemontoya replied:
Yes, but cautiously. By this I mean that we definitely need to put some tooling in place in the implementation to allow debugging and help operators know what is going on. However we also don't want to make the library implementing this bloated by too many extra tools and scaffolding.

nasthagiri reviewed (approved):
This looks great!  Thank you for leading us in an aligned direction via this PR, the discourse thread, and live discussions.

felipemontoya commented:
Thanks a lot @nasthagiri and @nizarmah. I updated the two suggestions you made and went ahead to make the status be Provisional.

Would agree to merge this in this Provisional state so that we can focus on the discussions at the libraries about naming and ergonomics? I'd like to merge so that this starts to appear at https://open-edx-proposals.readthedocs.io/en/latest/

nasthagiri commented:
Yes, go for it. I had approved the PR and we had kept it open for comments already.

edx-webhook commented:
@felipemontoya 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/open-edx-proposals&entry.867055334=https://github.com/edx/open-edx-proposals/pull/184&entry.1484655318=https://github.com/felipemontoya&entry.752974735=2021-02-05+00:00&entry.1917517419=2021-05-20+00:10&entry.2133058324=Yes) so we can improve your experience in the future.

edx-webhook commented:
@jmbowman, @feanil: thought you might like to know that felipemontoya merged this pull request.

<!-- comment:champion_merge_ping -->
